### PR TITLE
Implement RouteHandlerServices.Map and use in RequestDelegateGenerator

### PR DESCRIPTION
**Pull request by captainsafia**
Addresses https://github.com/dotnet/aspnetcore/issues/46163

- Adds a `RouteHandlerServices.Map` implementation per API review above
- Remove `SourceGeneratorEndpointDataSource` implementation in RequestDelegateGenerator's output
- Make `IEndpointRouteBuilder.GetOrAddRouteEndpointDataSource` internal to support invocation from `RouteHandlerServices.Map`
- Update generated thunks to match signature expected by `RouteHandlerServices.Map`

### CODE REVIEW

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs [hunk 1]
```diff
@@ -47,16 +47,17 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
 
         var thunks = endpoints.Select((endpoint, _) => $$"""
             [{{StaticRouteHandlerModelEmitter.EmitSourceKey(endpoint)}}] = (
-           (del, builder) =>
+           (methodInfo, options) =>
```
Comment (JamesNK) [inline src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:50]: What do you think of explicitly specifying the lambda types here and the build func?

Output from source generators I've seen is always very conservative. Explicit types might reduce the chances of any conflicts or ambiguous method calls in the future.

File: src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs [hunk 2]
```diff
@@ -174,239 +157,5 @@ internal static Task ExecuteObjectResult(object? obj, HttpContext httpContext)
         }
     }
 }
-
-{{GeneratedCodeAttribute}}
-file class SourceGeneratedRouteEndpointDataSource : EndpointDataSource
```
Comment (JamesNK) [inline src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs:179]: 🥇

File: src/Http/Routing/src/Builder/RouteHandlerServices.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    /// Registers an endpoint with custom functions for constructing
+    /// a request delegate for its handler and populating metadata for
+    /// the endpoint. Intended for consumption in the RequestDelegateGenerator.
+    /// </summary>
+    /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the route to.</param>
+    /// <param name="pattern">The route pattern.</param>
+    /// <param name="handler">The delegate executed when the endpoint is matched.</param>
+    /// <param name="httpMethods">The set of supported HTTP methods. May not be null.</param>
+    /// <param name="populateMetadata">A delegate for populating endpoint metadata.</param>
+    /// <param name="createRequestDelegate">A delegate for constructing a RequestDelegate from a Delegate.</param>
```
Comment (JamesNK) [inline src/Http/Routing/src/Builder/RouteHandlerServices.cs:31]:
```suggestion
    /// <param name="createRequestDelegate">A delegate for constructing a RequestDelegate.</param>
```

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs (hunk 1 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:50]: Is that really any different from any other public API we ship? The customer might be calling code and we might add overloads. I'm not convinced we should change anything.

Comment (JamesNK) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:50]: I'm thinking about the situation where the generator code is somehow used on a newer framework (no, I don't know how that could happen. Throwing out a what-if situation)

The difference is the customer controls their code and can fix it. If a generator outputs code that doesn't compile, then it's not immediately obvious what has gone wrong and what someone should do.

My point of view is generated code is always ugly. It's not like we lose anything by being explicit.

Comment (captainsafia) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:50]: I was motivated to avoid hardcoding too many things here under the assumption that the generated code is going to change quite a bit over the next preview. This aspect of it might not change but I valued making it easy to modify before we did any hardening to it. I think there's a lot of value in generated code that is easier to read especially as we spin things up.

File: src/Http/Routing/src/Builder/RouteHandlerServices.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+namespace Microsoft.AspNetCore.Routing;
+
+/// <summary>
+/// Provides methods used for invoking the route endpoint
+/// infrastructure with custom funcs for populating metadata
+/// and creating request delegates. Intended to be consumed from
+/// the RequestDeleatgeGenerator only.
+/// </summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+public static class RouteHandlerServices
```
Comment (Youssef1313) [inline src/Http/Routing/src/Builder/RouteHandlerServices.cs:19]: Can't this be added as a `file` class in the generated code, instead of being public that is not intended for developers to call?

Comment (davidfowl) [reply src/Http/Routing/src/Builder/RouteHandlerServices.cs:19]: No, there's logic that we want to keep in the framework.

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs [hunk 5]
```diff
@@ -47,16 +47,21 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
...
             {
-                builder.Metadata.Add(new SourceKey{{StaticRouteHandlerModelEmitter.EmitSourceKey(endpoint)}});
+                if (options == null)
+                {
+                    return new RequestDelegateMetadataResult { EndpointMetadata = new List<object>().AsReadOnly() };
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:54]: nit:
Replace `new List<object>().AsReadOnly()` with `ReadOnlyCollection<object>.Empty`, at least 3 places can be changed

Comment (captainsafia) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:54]: TIL! Looks like this was added in 8.0? Very nice...

Comment (BrennanConroy) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:54]: Oh, I just assumed it existed and saw that it did 😄 
Didn't realize it's new!

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs [hunk 6]
```diff
@@ -39,6 +39,40 @@ public async Task MapAction_NoParam_StringReturn(string source, string expectedB
         Assert.Equal(expectedBody, body);
     }
 
+    [Fact]
+    public async Task MapGet_NoParam_StringReturn_WithFilter()
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:43]: Not necessarily for this PR, but what do you think about having a small set of tests (1-5?) that baseline the generated code. That way it is easy in PRs to see what changes are being made to the generated code. In my opinion, it helps spot bugs if I can read the generated code.

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs (hunk 5 above)
Comment (eerhardt) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs:54]: I made the same comment, but you can use `Array.Empty<object>()` instead.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs (hunk 6 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.cs:43]: I went ahead and added the infrastructure to do this and applied it to this test.

File: src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs [hunk 7]
```diff
@@ -88,47 +88,30 @@ internal static class GenericThunks<T>
...
+        var (populateMetadata, createRequestDelegate) = GenericThunks<T>.map[(filePath, lineNumber)];
+        return RouteHandlerServices.Map(routes, pattern, handler, httpMethods, populateMetadata, createRequestDelegate);
     }
 
+    {{GeneratedCodeAttribute}}
     internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder MapCore(
         this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder routes,
         string pattern,
-        System.Delegate handler,
+        global::System.Delegate handler,
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs]: I think now is a good time to make this change before we fill out the generated code some more. I've updated the structure to use this technique. Let me know if [this looks right](https://github.com/dotnet/aspnetcore/blob/safia/map-rdsg/src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt).

File: src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs (hunk 5 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs]: I ended up using `ReadOnlyCollection<object>`. In the interest of readability it does a better job of communicating the type passed to the initializer here and I don't think there's any memory implications with that singleton.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+%GENERATEDCODEATTRIBUTE%
+internal static class GenerateRouteBuilderEndpoints
+{
+    private static readonly string[] GetVerb = new[] { HttpMethods.Get };
+    private static readonly string[] PostVerb = new[] { HttpMethods.Post };
+    private static readonly string[] PutVerb = new[]  { HttpMethods.Put };
+    private static readonly string[] DeleteVerb = new[] { HttpMethods.Delete };
+    private static readonly string[] PatchVerb = new[] { HttpMethods.Patch };
+
+    %GENERATEDCODEATTRIBUTE%
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:42]: Do we need the attribute here, if we already have it on the class?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    using Microsoft.AspNetCore.Routing;
+    using Microsoft.AspNetCore.Routing.Patterns;
+    using Microsoft.AspNetCore.Builder;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Extensions.FileProviders;
+    using Microsoft.Extensions.Primitives;
+
+    using MetadataPopulator = System.Func<System.Reflection.MethodInfo, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions?, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult>;
+    using RequestDelegateFactoryFunc = System.Func<System.Delegate, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult?, Microsoft.AspNetCore.Http.RequestDelegateResult>;
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:75]: These can be simplified now, for readability.

```suggestion
    using MetadataPopulator = Func<MethodInfo, RequestDelegateFactoryOptions?, RequestDelegateMetadataResult>;
    using RequestDelegateFactoryFunc = Func<Delegate, RequestDelegateFactoryOptions, RequestDelegateMetadataResult?, RequestDelegateResult>;
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+    %GENERATEDCODEATTRIBUTE%
+    file static class GeneratedRouteBuilderExtensionsCore
+    {
+        internal static class GenericThunks<T>
+        {
+            public static readonly System.Collections.Generic.Dictionary<(string, int), (MetadataPopulator, RequestDelegateFactoryFunc)> map = new()
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:82]:
```suggestion
            public static readonly Dictionary<(string, int), (MetadataPopulator, RequestDelegateFactoryFunc)> map = new()
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            {
+
+            };
+        }
+
+        internal static readonly System.Collections.Generic.Dictionary<(string, int), (MetadataPopulator, RequestDelegateFactoryFunc)> map = new()
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:88]:
```suggestion
        internal static readonly Dictionary<(string, int), (MetadataPopulator, RequestDelegateFactoryFunc)> map = new()
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+                if (options == null)
+                {
+                    return new RequestDelegateMetadataResult { EndpointMetadata = ReadOnlyCollection<object>.Empty };
+                }
+                options.EndpointBuilder.Metadata.Add(new SourceKey(@"TestMapActions.cs", 15));
+                return new RequestDelegateMetadataResult { EndpointMetadata = options.EndpointBuilder.Metadata.AsReadOnly() };
+            },
+            (del, options, inferredMetadataResult) =>
+            {
+                var handler = (System.Func<string>)del;
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:102]:
```suggestion
                var handler = (Func<string>)del;
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+                EndpointFilterDelegate? filteredInvocation = null;
+
+                if (options.EndpointBuilder.FilterFactories.Count > 0)
+                {
+                    filteredInvocation = GeneratedRouteBuilderExtensionsCore.BuildFilterDelegate(ic =>
+                    {
+                        if (ic.HttpContext.Response.StatusCode == 400)
+                        {
+                            return System.Threading.Tasks.ValueTask.FromResult<object?>(Results.Empty);
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:111]:
```suggestion
                            return ValueTask.FromResult<object?>(Results.Empty);
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+                        }
+                        return System.Threading.Tasks.ValueTask.FromResult<object?>(handler());
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:113]:
```suggestion
                        return ValueTask.FromResult<object?>(handler());
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+                    },
+                    options.EndpointBuilder,
+                    handler.Method);
+                }
+
+                System.Threading.Tasks.Task RequestHandler(Microsoft.AspNetCore.Http.HttpContext httpContext)
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:119]:
```suggestion
                Task RequestHandler(HttpContext httpContext)
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+                {
+                        var result = handler();
+                        return httpContext.Response.WriteAsync(result);
+                }
+                async System.Threading.Tasks.Task RequestHandlerFiltered(Microsoft.AspNetCore.Http.HttpContext httpContext)
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:124]:
```suggestion
                async Task RequestHandlerFiltered(HttpContext httpContext)
```

Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:124]: You get the point. :) I won't add more of these.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+                }
+
+                RequestDelegate targetDelegate = filteredInvocation is null ? RequestHandler : RequestHandlerFiltered;
+                var metadata = inferredMetadataResult?.EndpointMetadata ?? ReadOnlyCollection<object>.Empty;
+                return new RequestDelegateResult(targetDelegate, metadata);
+            }),
+
+        };
+
+        %GENERATEDCODEATTRIBUTE%
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:137]: Similar comment as above. Do we need this attribute on these methods when the whole class is marked as generated?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+        {
+            T.PopulateMetadata(method, builder);
+        }
+
+        internal static void PopulateMetadata<T>(System.Reflection.ParameterInfo parameter, EndpointBuilder builder) where T : Microsoft.AspNetCore.Http.Metadata.IEndpointParameterMetadataProvider
+        {
+            T.PopulateMetadata(parameter, builder);
+        }
+
+        internal static Task ExecuteObjectResult(object? obj, HttpContext httpContext)
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:190]: Do all of these methods need to be `internal`? Same for the `static class GenericThunks<T>` and the `static readonly Dictionary map`?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+    using Microsoft.Extensions.FileProviders;
+    using Microsoft.Extensions.Primitives;
+
+    using MetadataPopulator = System.Func<System.Reflection.MethodInfo, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions?, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult>;
+    using RequestDelegateFactoryFunc = System.Func<System.Delegate, Microsoft.AspNetCore.Http.RequestDelegateFactoryOptions, Microsoft.AspNetCore.Http.RequestDelegateMetadataResult?, Microsoft.AspNetCore.Http.RequestDelegateResult>;
+
+    %GENERATEDCODEATTRIBUTE%
+    file static class GeneratedRouteBuilderExtensionsCore
+    {
+        internal static class GenericThunks<T>
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:80]: Should we even generate this class if it is empty? It kind of looks weird. It also looks like a bug when you look at the `MapCore<T>` method that indexes into the empty dictionary in this class.

Seems like more code to compile, that isn't used.

**Review by eerhardt — COMMENTED** [review]
This is awesome work, @captainsafia.

I feel like I'm better able to give a decent code review now that I can see the generated code.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 9 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:75]: Not sure it can be. Last I tried this doesn't work, or the IDE was trolling me.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 19 above)
Comment (davidfowl) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:80]: Yea, this needs to be cleaned up.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 9 above)
Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:75]: I'd be really surprised if this doesn't work. Note that @captainsafia refactored the generated code to be inside a namespace with all the required usings inside the namespace declaration. With that work, this code can be written just like a "normal dev" would write code.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 12 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:102]: This is the only fully-qualified reference I wasn't able to remove. There's one function that generates the delegate definition that's used in both the body of the RequestDelegateGenerator and in the strongly-typed `MapAction` overload. Since the `MapAction` overload needs the fully qualified type we do that in both places. It's not a very nested type so hopefully this is fine.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 18 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: Nope, the only things that need to be internal are the `MapCore` methods. The fact that these were internal was an artifact of the previous model where we used partial classes that separated out the `MapAction` definitions from the rest of the generated code.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 19 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: Switched to only generated these if we produced thunks from the codegen.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 9 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:69]: It looks like fully-qualified namespaces are a requirement of type aliases regardless of what other usings exist ([ref](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-directive#using-alias)).

> Create a using alias directive to make it easier to qualify an identifier to a namespace or type. In any using directive, the fully qualified namespace or type must be used regardless of the using directives that come before it.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 17 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: I noodled on our use of the `GeneratedCode` attribute here a little bit more. Since the primary goal of the attribute is to make it easier to inspect if a type is codegenned from metadata, I think it makes sense to only include it in types that could be referenced from user code. Specifically, this would be:

- The `SourceKey` definition
- The strongly-typed `Map` overloads

I refactored the code to apply it to these places. Note, I opted to remove it from the static class containing the strongly-typed `Map` overloads since it's unlikely that user code ends up resolving that directly.

Let me know if you think this philosophy makes sense

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 12 above)
Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:88]: I think this is OK, I'm not too concerned about it.

If we did want to eliminate this, `StaticRouteHandlerModelEmitter.EmitHandlerDelegateType` could take a `bool fullyQualifyTypeNames` bool and optionally emit the `global::Namespace` part. But I think this is fine for now.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 17 above)
Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: I believe we want the `GeneratedCodeAttribute` on all the classes we generate in this file.

The reason is so things like analyzers, code-style, and code coverage tools don't warn/report things from inside this code. The devs can't change this code, so they shouldn't see tools talking about.

Even for the `internal static class GenerateRouteBuilderEndpoints` class. The Regex generator puts it on the partial methods, and not the class, because the class is partial to the user's code. So it would be incorrect to put it on the whole class, since the user writes code in this class too.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+{
+    %GENERATEDCODEATTRIBUTE%
+    internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder MapGet(
+        this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder endpoints,
+        [global::System.Diagnostics.CodeAnalysis.StringSyntax("Route")] string pattern,
+        global::System.Func<string> handler,
+        [global::System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
+        [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
+    {
+        return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(endpoints, pattern, handler, new[] { HttpMethods.Get }, filePath, lineNumber);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:43]: We should be, the original code did 😄

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 17 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: > The Regex generator puts it on the partial methods, and not the class, because the class is partial to the user's code. So it would be incorrect to put it on the whole class, since the user writes code in this class too.

This insights makes sense to me.

Here's the modification I made:

- Moved the `GenerateRouteBuilderEndpoints` to the code generated in the `Microsoft.AspNetCore.Builder` namespace to fix the `global::` using issue
- Added the `GeneratedCode` attribute to the `GenerateRouteBuilderEndpoints` class

So, all the classes we codegen inside that namespacce are marked.

Classes marked as `file private` are not marked, which I assume is fine.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: 🙊 🦭

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt (hunk 17 above)
Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt]: > Classes marked as file private are not marked, which I assume is fine.

I think they should be marked as well. That way they don't show up in code coverage reports, that devs can't do anything about.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 21]
```diff
@@ -7,10 +7,11 @@
 // </auto-generated>
 //------------------------------------------------------------------------------
 #nullable enable
-using global::Microsoft.AspNetCore.Http;
 
 namespace Microsoft.AspNetCore.Builder
 {
+    using Microsoft.AspNetCore.Http;
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:13]: Why is this necessary?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 22]
```diff
@@ -23,26 +24,33 @@ namespace Microsoft.AspNetCore.Builder
...
     {
-        return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(endpoints, pattern, handler, new[] { HttpMethods.Get }, filePath, lineNumber);
-    }
+        private static readonly string[] GetVerb = new[] { HttpMethods.Get };
+        private static readonly string[] PostVerb = new[] { HttpMethods.Post };
+        private static readonly string[] PutVerb = new[]  { HttpMethods.Put };
+        private static readonly string[] DeleteVerb = new[] { HttpMethods.Delete };
+        private static readonly string[] PatchVerb = new[] { HttpMethods.Patch };
+
+        %GENERATEDCODEATTRIBUTE%
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:42]: We shouldn't need the attribute here because the class has this attribute.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+
+    // This class needs to be internal so that the compiled application
+    // has access to the strongly-typed endpoint definitions that are
+    // generated by the compiler so that they will be favored by
+    // overload resolution and opt the runtime in to the code generated
+    // implementation produced here.
+    %GENERATEDCODEATTRIBUTE%
+    internal static class GenerateRouteBuilderEndpoints
+    {
+        private static readonly string[] GetVerb = new[] { HttpMethods.Get };
```
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:36]:
```suggestion
        private static readonly string[] GetVerb = new[] { global::Microsoft.AspNetCore.Http.HttpMethods.Get };
```

Comment (eerhardt) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:36]: Everything else here is globally qualified.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for fixing all my nits. :) I think the code gen looks really clean now.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+
+        private static void PopulateMetadata<T>(MethodInfo method, EndpointBuilder builder) where T : IEndpointMetadataProvider
+        {
+            T.PopulateMetadata(method, builder);
+        }
+
+        private static void PopulateMetadata<T>(ParameterInfo parameter, EndpointBuilder builder) where T : IEndpointParameterMetadataProvider
+        {
+            T.PopulateMetadata(parameter, builder);
+        }
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapGet_NoParam_StringReturn_WithFilter.generated.txt:165]: Yes 🙊 ... we'll eventually use this when we add support for populating metadata for types that implement the `IEndpointMetadataProvider` interfaces. I can remove it in the next PR...
